linkerd/app/core/src/control.rs

```diff
@@ -145,160 +145,375 @@ pub mod add_origin {
+                Ok(dns_rsp(vec!["127.0.0.1"], Duration::from_millis(300))),
+            ]);
+
+            let mut stream = task::spawn(
+                resolution_stream(dns_resolver, address("test.com:8888"))
+                    .map_err(|err| err.to_string()),
+            );
+            assert_ready_eq!(
+                stream.poll_next(),
+                Some(Ok(add(vec![
+                    "127.0.0.1:8888",
+                    "127.0.0.2:8888",
+                    "127.0.0.3:8888",
+                    "127.0.0.4:8888"
+                ])))
+            );
+            time::advance(Duration::from_millis(101)).await;
+            assert_ready_eq!(stream.poll_next(), Some(Ok(Update::DoesNotExist)));
+            time::advance(Duration::from_millis(101)).await;
+            assert_ready_eq!(stream.poll_next(), Some(Ok(add(vec!["127.0.0.1:8888"]))));
+        }
+
+        #[tokio::test]
+        async fn yields_empty() {
+            time::pause();
+            let dns_resolver = MockDnsResolver::new(vec![
+                Ok(dns_rsp(
+                    vec!["127.0.0.1", "127.0.0.2", "127.0.0.3", "127.0.0.4"],
+                    Duration::from_millis(100),
+                )),
+                Err(dns::Error::NoAddressesFound(
+                    Instant::now() + Duration::from_millis(200),
+                    true,
+                )),
+                Ok(dns_rsp(vec!["127.0.0.1"], Duration::from_millis(300))),
+            ]);
+
+            let mut stream = task::spawn(
+                resolution_stream(dns_resolver, address("test.com:8888"))
+                    .map_err(|err| err.to_string()),
+            );
+            assert_ready_eq!(
+                stream.poll_next(),
+                Some(Ok(add(vec![
+                    "127.0.0.1:8888",
+                    "127.0.0.2:8888",
+                    "127.0.0.3:8888",
+                    "127.0.0.4:8888"
+                ])))
+            );
+            time::advance(Duration::from_millis(101)).await;
+            assert_ready_eq!(stream.poll_next(), Some(Ok(Update::Empty)));
+            time::advance(Duration::from_millis(101)).await;
+            assert_ready_eq!(stream.poll_next(), Some(Ok(add(vec!["127.0.0.1:8888"]))));
+        }
+
+        #[tokio::test]
+        #[ignore]
+        async fn continues_yielding_after_err() {
+            time::pause();
+            let dns_resolver = MockDnsResolver::new(vec![
+                Err(dns::Error::TaskLost),
+                Ok(dns_rsp(vec!["127.0.0.1"], Duration::from_millis(100))),
+            ]);
+
+            let mut stream = task::spawn(
+                resolution_stream(dns_resolver, address("test.com:8888"))
+                    .map_err(|err| err.to_string()),
+            );
+            let result = assert_ready!(stream.poll_next());
+            assert!(result.unwrap().is_err());
+            assert_ready_eq!(stream.poll_next(), Some(Ok(add(vec!["127.0.0.1:8888"]))));
+        }
```

Comment:
This should really keep yielding after error. Need to verify the assumptions async-stream makes.